### DISCUSSION TITLE
TCL as possible compile target

### DISCUSSION THREAD

**Discussion by helegrod** [fable-compiler/Fable#3196] [Ideas]
Hi everyone. I'm interested in investigating TCL as a compile target for Fable. 

TCL is used extensively in the hardware design industry, as it's the only language a lot of our tooling understands. We end up writing very long, difficult to maintain scripts that would greatly benefit from the type-safety and rigor of F# and its toolchain. 

Where would I get started exploring this? I've used F# and Fable for a while, but I've never done much work with compilers or looked at the internals of Fable.

The main site for the language is linked below.
[https://www.tcl.tk/](https://www.tcl.tk/)

**Comment by alfonsogarciacaro** [comment]
Thanks for your interesting in expanding Fable @BrothersRM!

Unfortunately, we haven't abstracted yet the language target in the compiler, so there are a few places you need to wire manually to add a new language, but the compiler should help you along the way.

These are the modules you need to create to add a new language:

- **MyLang.fs**: The AST for your language
- **Fable2MyLang**: Transform Fable AST into your language AST
- **MyLangPrinter**: Print your language AST
- **MyLangReplacements**: Tell Fable how calls to FSharp.Core or .NET BCL should be replaced

These are usually in a language-specific folder within `src/Fable.Transforms`, although the Replacements file needs to appear on top of Fable.Transforms.fsproj because of dependencies.

What I would do is take one of the languages as base and transform it until it looks like TCL :) I would suggest Dart. You can try to start duplicating all Dart files and whenever there's a language switch, as in [src/Fable.Cli/Pipeline.fs](https://github.com/fable-compiler/Fable/blob/9daa6d98a016d9374f08de8d400fa6a260ce8e15/src/Fable.Cli/Pipeline.fs#L343-L349). Or to make it easier, in your experimental branch you can just replace Dart with TCL incrementally. We can add Dart files later.

Not sure how's the testing story with TCL but I will also try to setup a minimal test suite quickly so you can progress with confidence.

**Reply by helegrod** [reply]
That's very interesting, thank you for the guidance! I'll spend some time reading over the code to get a sense of how things work.
